File: csharp/src/Utils/rest.cs
//using System.Net.Http;
using System.Net.Http.Headers;
// for RSA
using System.Security.Cryptography;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using StringDict = System.Collections.Generic.Dictionary<string, string>;

/// <summary>
/// Generic REST client with basic and oauth.
/// Replace with your best REST client object
/// or use openapi generator to generate stubs.
/// </summary>
public class Rest
{
    // take come time back to account for time offset between client and server
    private const int JWT_NOT_BEFORE_OFFSET_SEC = 60;
    // take some validity for the JWT
    private const int JWT_EXPIRY_OFFSET_SEC = 600;
    private StringDict mApiData;
    private Rest mOAuthAPI;
    private HttpClient mHttpClient;

    /// <summary>
    /// Read RSA private key from file.
    /// </summary>
    /// <param name="filename"></param>
    /// <returns>RSA key</returns>
    /// <exception cref="InvalidOperationException"></exception>
    private static RSA readKeyFromFile(string filename)
    {
        string pemContents = System.IO.File.ReadAllText(filename);
        const string RsaPrivateKeyHeader = "-----BEGIN RSA PRIVATE KEY-----";
        const string RsaPrivateKeyFooter = "-----END RSA PRIVATE KEY-----";

        if (!pemContents.StartsWith(RsaPrivateKeyHeader))
        {
            throw new InvalidOperationException();
        }
        int endIdx = pemContents.IndexOf(
            RsaPrivateKeyFooter,
            RsaPrivateKeyHeader.Length,
            StringComparison.Ordinal);

        string base64 = pemContents.Substring(
            RsaPrivateKeyHeader.Length,
            endIdx - RsaPrivateKeyHeader.Length);

        byte[] der = Convert.FromBase64String(base64);
        RSA rsa = RSA.Create();
        rsa.ImportRSAPrivateKey(der, out _);
        return rsa;
    }
    public Rest(StringDict api_data)
    {
        // shallow copy sufficient here
        mApiData = api_data.ToDictionary(entry => entry.Key, entry => entry.Value);

        mHttpClient = new HttpClient
        {
            BaseAddress = new Uri(mApiData["base_url"])
        };
    }
    public string get_bearer_token(string scope)
    {
        long seconds_since_epoch = System.DateTimeOffset.Now.ToUnixTimeSeconds();
        var payload = new JObject{
                    { "iss", mApiData["oauth_client_id"]},
                    { "sub", mApiData["oauth_jwt_subject"]},
                    { "aud", mApiData["oauth_jwt_audience"]},
                    { "nbf", seconds_since_epoch - JWT_NOT_BEFORE_OFFSET_SEC},
                    { "iat", seconds_since_epoch - JWT_NOT_BEFORE_OFFSET_SEC},
                    { "exp", seconds_since_epoch + JWT_EXPIRY_OFFSET_SEC},
                };
        // if client id starts with "aspera", add key "org" to payload
        if (mApiData["oauth_client_id"].StartsWith("aspera."))
        {
            payload["org"] = mApiData["aoc_org"];
        }
        Log.DumpJObject("payload", payload);
        var private_key = readKeyFromFile(mApiData["oauth_file_private_key"]);
        string assertion = Jose.JWT.Encode(JsonConvert.SerializeObject(payload), private_key, Jose.JwsAlgorithm.RS256, extraHeaders: new Dictionary<string, object> { { "typ", "JWT" } });
        var token_params = new JObject{
            {"www_body_params",true},
            {"client_id",mApiData["oauth_client_id"]},
            {"grant_type","urn:ietf:params:oauth:grant-type:jwt-bearer"},
            {"assertion",assertion},
        };
        if (scope != null)
        {
            token_params["scope"] = scope;
        }
        JObject data = (JObject)mOAuthAPI.create(mApiData["oauth_path_token"], token_params);
        return (string)data["access_token"];
    }
    public string get_bearer(string scope)
    {
        return $"Bearer {get_bearer_token(scope)}";
    }
    /// <summary>
    /// Call REST API.
    /// </summary>
    /// <param name="operation">GET, ...</param>
    /// <param name="subpath">endpoint</param>
    /// <param name="json_params"></param>
    /// <param name="url_params"></param>
    /// <param name="headers"></param>
    /// <returns></returns>
    /// <exception cref="System.Exception"></exception>
    public JContainer call(HttpMethod operation, string subpath, JObject json_params = null, JObject url_params = null, StringDict headers = null)
    {
        string uri_string = mApiData["base_url"] + "/" + subpath;
        var builder = new System.UriBuilder(uri_string);// { Query = collection.ToString() };
        if (url_params != null)
        {
            var query = url_params.Properties().ToDictionary(p => p.Name, p => p.Value.ToString());
            string query_string = "";
            foreach (var key in query.Keys)
            {
                if (query_string.Length != 0)
                {
                    query_string = query_string + "&";
                }
                query_string = query_string + System.Uri.EscapeDataString(key) + "=" + System.Uri.EscapeDataString("" + query[key]);
            }
            builder.Query = query_string;
        }
        HttpRequestMessage request = new HttpRequestMessage
        {
            Method = operation,
            RequestUri = builder.Uri
        };
        switch (mApiData["type"])
        {
            case "oauth2":
                mOAuthAPI = new Rest(new StringDict(){
                    {"base_url",mApiData["oauth_base_url"]},
                    {"type","basic"},
                    {"basic_username",mApiData["oauth_client_id"]},
                    {"basic_password",mApiData["oauth_client_secret"]},
                });
                string scope = null;
                if (mApiData.ContainsKey("oauth_scope"))
                {
                    scope = mApiData["oauth_scope"];
                }
                request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", get_bearer_token(scope));
                break;
            case "basic":
                var encoded = Convert.ToBase64String(System.Text.ASCIIEncoding.ASCII.GetBytes($"{mApiData["basic_username"]}:{mApiData["basic_password"]}"));
                request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Basic", encoded);
                break;
            default:
                throw new System.Exception("wrong auth type");
        }
        if (json_params != null)
        {
            Log.log.Debug($"json_params {json_params}");
            if (json_params.ContainsKey("www_body_params"))
            {
                json_params.Remove("www_body_params");
                request.Content = new FormUrlEncodedContent(json_params.Properties().ToDictionary(p => p.Name, p => p.Value.ToString()));
            }
            else
            {
                request.Content = new StringContent(
                    JsonConvert.SerializeObject(json_params),
                    System.Text.Encoding.UTF8,
                    "application/json");
            }
        }
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        Log.DumpJObject("req", request);
        if (request.Content != null)
        {
            Log.log.Debug($"data={request.Content.ReadAsStringAsync().Result}");
        }
        var response = mHttpClient.SendAsync(request).Result;
        var resp_str = response.Content.ReadAsStringAsync().Result;
        Log.log.Debug($"resp={resp_str}");
        if (!response.IsSuccessStatusCode)
        {
            throw new System.Exception($"ERROR: {response.StatusCode} {response.ReasonPhrase}");
        }
        JContainer result = null;
        if (resp_str.Length != 0)
        {
            if (resp_str.StartsWith("["))
            {
                result = JArray.Parse(resp_str);
            }
            else
            {
                result = JObject.Parse(resp_str);
            }
        }
        return result;
    }
    public JContainer create(string subpath, JObject json_params)
    {
        return call(operation: HttpMethod.Post, subpath: subpath, headers: new StringDict { { "Accept", "application/json" } }, json_params: json_params);
    }
    public JContainer read(string subpath, JObject url_params = null)
    {
        return call(operation: HttpMethod.Get, subpath: subpath, headers: new StringDict { { "Accept", "application/json" } }, url_params: url_params);
    }
    public JContainer update(string subpath, JObject json_params)
    {
        return call(operation: HttpMethod.Put, subpath: subpath, headers: new StringDict { { "Accept", "application/json" } }, json_params: json_params);
    }
    public JContainer delete(string subpath)
    {
        return call(operation: HttpMethod.Delete, subpath: subpath, headers: new StringDict { { "Accept", "application/json" } });
    }
}

File: rust/src/utils/rest.rs
// cspell:ignore reqwest jsonwebtoken
use serde_json::Value;
use std::collections::HashMap;
use std::error::Error;
use std::time::{SystemTime, UNIX_EPOCH};

pub const MIME_JSON: &str = "application/json";
pub const MIME_FORM: &str = "application/x-www-form-urlencoded";
pub const IETF_GRANT_JWT: &str = "urn:ietf:params:oauth:grant-type:jwt-bearer";

// Offset allowed between client and server
const JWT_NOT_BEFORE_OFFSET_SEC: usize = 60;
// Validity period for JW Token
const JWT_EXPIRY_OFFSET_SEC: usize = 600;

/// Information needed to generate a bearer token
#[derive(Clone)]
pub struct BearerData {
    pub token_url: String,
    pub key_pem_path: String,
    pub client_id: String,
    pub client_secret: String,
    pub iss: String,
    pub aud: String,
    pub sub: String,
    pub org: Option<String>,
}
/// Information needed to generate a basic token
pub struct BasicData {
    pub username: String,
    pub password: String,
}
/// Enum to store the authentication data
pub enum AuthData {
    Bearer(BearerData),
    Basic(BasicData),
    None,
}
/// JWT Claims structure
#[derive(Debug, serde::Serialize, serde::Deserialize)]
struct Claims {
    iss: String,
    aud: String,
    sub: String,
    exp: usize,
    nbf: usize,
    iat: usize,
    jti: String,
    org: Option<String>,
}
/// REST API client
pub struct Rest {
    base_url: String,
    auth: AuthData,
    headers: HashMap<String, String>,
    client: reqwest::Client,
}
impl Rest {
    /// Create a new REST API client
    ///
    /// ### Arguments
    /// * `base_url` - The base URL for the API
    /// * `verify` - Whether to verify SSL certificates
    pub fn new(base_url: &str, verify: bool) -> Result<Self, Box<dyn Error>> {
        let mut client_builder = reqwest::Client::builder();
        if !verify {
            client_builder = client_builder.danger_accept_invalid_certs(true);
        }
        Ok(Self {
            base_url: base_url.to_string(),
            auth: AuthData::None,
            headers: HashMap::new(),
            client: client_builder.build()?,
        })
    }
    /// API is authenticated using basic auth.
    ///
    /// ### Arguments
    /// * `username` - The username
    /// * `password` - The password
    pub fn set_basic(&mut self, username: &str, password: &str) {
        self.auth = AuthData::Basic(BasicData {
            username: username.to_owned(),
            password: password.to_owned(),
        });
    }
    /// API is authenticated using a bearer token.
    ///
    /// ### Arguments
    /// * `auth_data` - Information for JWT generation
    pub fn set_bearer(&mut self, auth_data: BearerData) {
        self.auth = AuthData::Bearer(auth_data);
    }

    /// Set the default scope for the bearer token and update the headers
    ///
    /// ### Arguments
    /// * `scope` - The scope to set
    pub async fn set_default_scope(&mut self, scope: Option<String>) -> Result<(), Box<dyn Error>> {
        let token = self.get_bearer_token(scope).await?;
        self.headers.insert("Authorization".to_string(), token);
        Ok(())
    }

    /// Get a bearer token from the server
    ///
    /// ### Arguments
    /// * `scope` - The scope to set
    pub async fn get_bearer_token(
        &mut self,
        scope: Option<String>,
    ) -> Result<String, Box<dyn Error>> {
        let now = SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs() as usize;
        // get copy of self.auth as BearerData, else return error
        let auth = match &self.auth {
            AuthData::Bearer(auth) => auth.clone(),
            _ => return Err(anyhow::anyhow!("Bearer").into()),
        };
        let claims = Claims {
            iss: auth.iss.to_owned(),
            sub: auth.sub.to_owned(),
            aud: auth.aud.to_owned(),
            iat: now - JWT_NOT_BEFORE_OFFSET_SEC,
            nbf: now - JWT_NOT_BEFORE_OFFSET_SEC,
            exp: now + JWT_EXPIRY_OFFSET_SEC,
            jti: uuid::Uuid::new_v4().to_string(),
            org: auth.org.to_owned(),
        };
        let private_key = std::fs::read_to_string(auth.key_pem_path)?;
        let assertion = jsonwebtoken::encode(
            &jsonwebtoken::Header::new(jsonwebtoken::Algorithm::RS256),
            &claims,
            &jsonwebtoken::EncodingKey::from_rsa_pem(private_key.as_bytes())?,
        )?;
        let mut data = vec![
            ("grant_type", IETF_GRANT_JWT.to_string()),
            ("client_id", auth.client_id.to_owned()),
            ("assertion", assertion),
        ];
        if let Some(scope) = scope {
            data.push(("scope", scope));
        }
        // debug data
        log::debug!("Bearer data: {:?}", data);
        let response = self
            .client
            .post(auth.token_url) // "http://localhost:12345")//
            .basic_auth(
                auth.client_id.to_owned(),
                Some(auth.client_secret.to_owned()),
            )
            .header("Accept", MIME_JSON)
            .header("Content-Type", MIME_FORM)
            .form(&data)
            .send()
            .await?;
        // check response error
        if !response.status().is_success() {
            return Err(
                anyhow::anyhow!("Failed to get access token: {}", "response.status()").into(),
            );
        }
        let jdata: Value = response.json().await?;
        let token = jdata["access_token"]
            .as_str()
            .ok_or_else(|| anyhow::anyhow!("Failed to get access token from response"))?;
        log::debug!("Bearer token: {:?}", token);
        Ok(format!("Bearer {token}"))
    }

    /// CRUD: Create
    ///
    /// ### Arguments
    /// * `path` - The path to the API endpoint
    /// * `value` - The JSON value to send
    /// * `query` - Optional query parameters
    pub async fn call(
        &self,
        method: reqwest::Method,
        path: &str,
        value: Option<&Value>,
        query: Option<&[(&str, &str)]>,
    ) -> Result<Option<Value>, Box<dyn Error>> {
        let mut request_builder: reqwest::RequestBuilder = self
            .client
            .request(method, &format!("{}/{path}", self.base_url))
            .header("Content-Type", MIME_JSON)
            .header("Accept", MIME_JSON);
        // loop on headers and add them to the request
        for (key, value) in &self.headers {
            request_builder = request_builder.header(key, value);
        }
        if let Some(query) = query {
            request_builder = request_builder.query(query);
        }
        // add basic if here
        if let AuthData::Basic(data) = &self.auth {
            request_builder = request_builder.basic_auth(&data.username, Some(&data.password));
        }
        // add json value if present
        if let Some(value) = value {
            request_builder = request_builder.json(value);
        }
        let response = request_builder.send().await?;
        // check http code and transform to error
        if !response.status().is_success() {
            log::debug!("response: {:?}", response);
            log::debug!("response.text: {:?}", response.text().await?);
            return Err(anyhow::anyhow!("Failed to create: {}", "response.status()").into());
        }
        match response.json().await {
            Ok(value) => Ok(Some(value)),
            Err(_) => Ok(None),
        }
    }
    pub async fn create(
        &self,
        path: &str,
        value: &Value,
        query: Option<&[(&str, &str)]>,
    ) -> Result<Value, Box<dyn Error>> {
        Ok(self
            .call(reqwest::Method::POST, path, Some(value), query)
            .await?
            .unwrap())
    }
    pub async fn read(
        &self,
        path: &str,
        query: Option<&[(&str, &str)]>,
    ) -> Result<Value, Box<dyn Error>> {
        Ok(self
            .call(reqwest::Method::GET, path, None, query)
            .await?
            .unwrap())
    }
    pub async fn update(&self, path: &str, value: &Value) -> Result<(), Box<dyn Error>> {
        let _ = self
            .call(reqwest::Method::PUT, path, Some(value), None)
            .await?;
        Ok(())
    }
    pub async fn delete(&self, path: &str) -> Result<(), Box<dyn Error>> {
        let _ = self.call(reqwest::Method::DELETE, path, None, None).await?;
        Ok(())
    }
}
